### PR TITLE
[FLINK-36964] Fix potential exception when SchemaChange in parallel w…

### PR DESCRIPTION
**Pull request by lvyanquan**
Fix potential exception when SchemaChange in parallel with Paimon Sink.

This close FLINK-36964 and FLINK-35888.

### CODE REVIEW

**Comment by yuxiqian** [comment]
Hi @lvyanquan, I have some concern about how BucketAssignOperator works with schema evolution strategy.

1) FlushEvent now works as a "pipeline drain indicator", which means after all data sink writer acknowledges them, there should't be any unhandled / uncommitted events flowing the whole pipeline, so SchemaRegistry could evolve downstream DB safely.

Normal Flush Event Semantic

However, the bucket assigning strategy might break that assumption, where data sink writers might receive data change events with stale schema, even after external schema evolution processes have finished.

Questionable FlushEvent Semantic

2) Potential schema operator hanging risks with "distributed" topology. 

Hanging risks

Basically same as described in #3680. In short, if a broadcast / custom partitioning topology is applied, then blocking one upstream partition will implicitly block all downstream partitions from handling events.

3) Why we need another Bucket assigner operator?

AFAIK all data change events have been hashed & distributed in PartitionOperator. Since adding a BucketAssignOperator does not change the parallelism, is there any reason we can't calculate the correct bucket partition ID in advance, instead of creating another partitioning topology?

**Comment by yuxiqian** [comment]
A little off topic, but we're really lacking E2e tests to real-life data sinks with complicated writing topologies. #3491 with higher pressure might be necessary to expose more issues like this in the future.

**Comment by lvyanquan** [comment]
Hi @yuxiqian, could you help to review this?

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/utils/SchemaUtils.java [hunk 1]
```diff
@@ -101,6 +101,9 @@ public static Schema applySchemaChangeEvent(Schema schema, SchemaChangeEvent eve
     private static Schema applyAddColumnEvent(AddColumnEvent event, Schema oldSchema) {
         LinkedList<Column> columns = new LinkedList<>(oldSchema.getColumns());
         for (AddColumnEvent.ColumnWithPosition columnWithPosition : event.getAddedColumns()) {
+            if (oldSchema.getColumnNames().contains(columnWithPosition.getAddColumn().getName())) {
+                continue;
+            }
```
Comment (yuxiqian) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/utils/SchemaUtils.java:106]: Seems we don't need this check after `SchemaUtils#isSchemaChangeEventRedundant` has guarded us from duplicately applying non-idempotent schema change events.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/PaimonEventSink.java [hunk 2]
```diff
@@ -63,7 +64,12 @@ public DataStream<Event> addPreWriteTopology(DataStream<Event> dataStream) {
                 // All Events after BucketAssignOperator are decorated with BucketWrapper.
                 .partitionCustom(
                         (bucket, numPartitions) -> bucket % numPartitions,
-                        (event) -> ((BucketWrapper) event).getBucket());
+                        (event) -> ((BucketWrapper) event).getBucket())
+                // Avoid disorder of FlushEvent and DataChangeEvent.
+                .transform(
+                        "RemoveDuplicateEvents",
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/PaimonEventSink.java:70]: Make operatorName consistent with class name

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/PreCommitOperator.java [hunk 3]
```diff
@@ -50,8 +69,17 @@ public void open() throws Exception {
 
     @Override
     public void processElement(StreamRecord<CommittableMessage<MultiTableCommittable>> element) {
+        if (catalog == null) {
+            this.catalog = FlinkCatalogFactory.createPaimonCatalog(catalogOptions);
+            // flinkMetricGroup could be passed after FLIP-371.
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/PreCommitOperator.java:74]: Our baseline is Flink 1.19 where FLIP-371 has been resolved. Can we remove this TODO now?

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/PreCommitOperator.java [hunk 4]
```diff
@@ -64,34 +92,44 @@ public void finish() {
...
+        super.snapshotState(context);
+        long checkpointId = context.getCheckpointId();
+        if (!multiTableCommittables.isEmpty()) {
+            multiTableCommittables.forEach(
+                    (multiTableCommittable) ->
+                            LOGGER.debug(
+                                    "Try to commit: "
+                                            + multiTableCommittable
+                                            + " in checkpoint "
+                                            + checkpointId));
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/PreCommitOperator.java:119]:
```suggestion
LOGGER.debug("Try to commit: {} in checkpoint {}", ...);
```

Thus, we don't need to pay for string concatenation with higher log level.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/bucket/BucketAssignOperator.java [hunk 5]
```diff
@@ -183,19 +189,24 @@ public void processElement(StreamRecord<Event> streamRecord) throws Exception {
...
             schemaMaps.put(schemaChangeEvent.tableId(), new TableSchemaInfo(schema, zoneId));
-            output.collect(
-                    new StreamRecord<>(
-                            new BucketWrapperChangeEvent(currentTaskNumber, (ChangeEvent) event)));
+            // Broadcast SchemachangeEvent.
+            for (int index = 0; index < totalTasksNumber; index++) {
+                output.collect(
+                        new StreamRecord<>(
+                                new BucketWrapperChangeEvent(index, (ChangeEvent) event)));
+            }
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/bucket/BucketAssignOperator.java:209]: Seems we can merge `event instanceof CreateTableEvent` and `event instanceof SchemaChangeEvent`?

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/bucket/BucketAssignOperator.java [hunk 6]
```diff
@@ -210,6 +221,7 @@ public void processElement(StreamRecord<Event> streamRecord) throws Exception {
         }
         long targetRowNum = table.coreOptions().dynamicBucketTargetRowNum();
         Integer numAssigners = table.coreOptions().dynamicBucketInitialBuckets();
+        LOGGER.debug("Succeed to get table info " + table);
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/bucket/BucketAssignOperator.java:224]:
```suggestion
        LOGGER.debug("Succeed to get table info {}", table);
```

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/bucket/FlushEventAlignmentOperator.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+import org.apache.flink.streaming.api.operators.ChainingStrategy;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/** Remove duplicate {@link FlushEvent} created by {@link BucketAssignOperator}. */
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/bucket/FlushEventAlignmentOperator.java:33]:
```suggestion
/** Align {@link FlushEvent}s broadcasted by {@link BucketAssignOperator}. */
```

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/bucket/FlushEventAlignmentOperator.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                LOG.info(currentSubTaskId + " send FlushEvent of " + sourceSubTaskId);
+                output.collect(new StreamRecord<>(new FlushEvent(sourceSubTaskId)));
+                sourceTaskIdToAssignBucketSubTaskIds.remove(sourceSubTaskId);
+            } else {
+                LOG.info(
+                        currentSubTaskId
+                                + " collect FlushEvent of "
+                                + sourceSubTaskId
+                                + " with subtask "
+                                + subtaskId);
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/bucket/FlushEventAlignmentOperator.java:81]: Ditto, avoid eager string concatenation

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/bucket/FlushEventAlignmentOperator.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+public class FlushEventAlignmentOperator extends AbstractStreamOperator<Event>
+        implements OneInputStreamOperator<Event, Event> {
+
+    private transient int totalTasksNumber;
+
+    /**
+     * Key: subtask id of {@link SchemaOperator}, Value: subtask ids of {@link
+     * BucketAssignOperator}.
+     */
+    private transient Map<Integer, Set<Integer>> sourceTaskIdToAssignBucketSubTaskIds;
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/bucket/FlushEventAlignmentOperator.java:43]: I'm not sure if it's enough to distinguish `FlushEvent`s?

Notice that each upstream partition might emit `FlushEvent`s individually, and all of them might be broadcast again in the `BucketAssignOperator`.

**Review by yuxiqian — CHANGES_REQUESTED** [review]
Thanks for @lvyanquan's great work, just left some comments.

File: flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/bucket/FlushEventAlignmentOperator.java (hunk 9 above)
Comment (lvyanquan) [reply flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/bucket/FlushEventAlignmentOperator.java:43]: Each `FlushEventAlignmentOperator` will receive `FlushEvent`s with the same `sourceTaskId` of upstream partition  from `BucketAssignOperator`, each upstream partition was distinguished by `sourceTaskId`.

Comment (yuxiqian) [reply flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-paimon/src/main/java/org/apache/flink/cdc/connectors/paimon/sink/v2/bucket/FlushEventAlignmentOperator.java:43]: That makes sense to me.

**Review by yuxiqian — APPROVED** [review]
It's nice to have corresponding E2e cases to verify changes in this PR. Could @leonardBang @ruanhang1993 please trigger the CI workflow?

**Comment by lvyanquan** [comment]
Last CI passed and rebase master to fix the conflict.

**Comment by leonardBang** [comment]
@lvyanquan Would you like to rebase to latest master branch as conflicts happens?

**Comment by lvyanquan** [comment]
Rebased to master.

**Review by leonardBang — APPROVED** [review]
Thanks @lvyanquan for the contribution and @yuxiqian for the review, LGTM
